ボウリングスコア計算プログラムを作成

samovar1214 opened the pull request.

samovar1214 committed:
ボウリングスコア計算プログラムを追加

yuuu commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,53 @@
+scores = score.split(',')
+
+shots = []
+scores.each do |s|
```
ここは `Array#map` と三項演算子を使うともっとスッキリ書けそうです。

yuuu commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,53 @@
+    end
+end
+
+point +=
```
9フレーム目の処理ですが、1〜8フレーム目の処理とほぼ同じになっています。
処理をまとめてしまって、挙動を変えたいところだけif式で分岐させたほうが、全体のコード量が少なくなり、読みやすくもなると思いますがどうでしょう？

yuuu commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,53 @@
+  two_frames = frames[n + 1] + frames[n + 2]
+  point +=
+    if frames[n][0] == 10
+      10 + two_frames[0] + two_frames[1]
```
ここの `10` は `frames[n].sum` と同義ですね。これは次の分岐についても同様です。
つまり、ここのif-elsif-else式はどの分岐を通っても `frames[n].sum` を加算している状況です。重複しているコードを分岐の前後に移動させることで、重複を解消できないか検討をお願いします。

yuuu replied:
`two_frames[0] + two_frames[1]` の部分は `two_frames[0..1].sum` と書くとコードの重複を減らせます👍️

yuuu commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,53 @@
+end
+
+point = 0
+(0..7).each do |n|
```
ここは `Enumerable#sum` を使うと次のように書くことができ、 `point` への加算を自分で行う必要がなくなります。

```ruby
point = (0..7).each.sum do |n|
  # ブロックの戻り値を各フレームのスコアにする
end
```

yuuu replied:
ブロックが1行で終わる場合は1文字の引数を使うこともありますが、このように複数行にわたるブロックで引数が1文字だと、その変数に格納されている値の意味がわかりづらくなります。
わかりやすい変数名に見直しをお願いします🙏

yuuu commented on 03.bowling/bowling.rb:
```diff
@@ -0,0 +1,53 @@
+
+frames = []
+one_frame = []
+shots.each do |s|
```
ここは `Enumerable#each_with_object` を使うと `frames = []` の行を省略でき、少しだけコードをスッキリさせられそうです 👍 
https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/each_with_object.html

samovar1214 replied:
詳しくご教示いただきありがとうございます～！✨
コメントいただいた以下の点を修正してみます💪🏻

- コマンドライン引数の分割・数値化を`Array#map`と三項演算子で書く
- `Enumerable#each_with_object`を使って`frames = []`を省略する
- `Enumerable#sum`を使って`point`の加算をする
- `frames[n].sum`の重複を解消する
- `two_frames[0] + two_frames[1]`の部分を`two_frames[0..1].sum`と書く
- 1～8フレーム目と9フレーム目の処理をまとめる
- ブロックの引数の変数名を分かりやすくする

samovar1214 committed:
変数名とpointの計算を修正

1球毎に分割し数字に変換するまでの処理を短い書き方に、フレーム毎に分割する部分はframes = []を省略できる書き方に修正しました。また、ブロックの引数の変数名を1文字の名前から分かりやすい名前に修正しました。

pointの計算はEnumerable#sumを使う形に変更しました。frames[n].sumの重複を解消し、1～8フレーム目と9フレーム目の処理をまとめました。
